### PR TITLE
[WIP|Discussion] Better socket polling (+ StreamPeer refactor)

### PR DESCRIPTION
**Pull request by Faless**
This is a fix for #4157 - Detect server clean close in TCP/SSL (the SSL part need some extra work to implement proper client close notify which is now detected as an error instead of a clean close, but still detected).
The first commit is the uglier version (no refactoring), just get the thing done.

Discussion
---
The second commit tries to do that better, but also changes the API a bit, and admittedly over-generalize.
The idea is that most StreamPeers do have a status and need to be polled to update it.
Polling can be, sometimes, an expensive operation. For this reason, instead of "polling when you ask the status" it would make sense to move to a more expressive approach where you can decide when to poll, and `get_status` will not poll the connection (like it use to do in TCP).
Also, in the effort of diminishing duplicate status enums, and make them more consistent, I tried to have a simpler generalized StreamStatus (CLOSED-CONNECTING/OPENING-CONNECTED/OPENED/ERROR). Additional, stream-specific status could be added to the specific stream (e.g `StreamPeerSSL.get_ssl_error()` to get the SSL-specific error when `get_status` return `ERROR`).

Being honest, I'm not yet fully convinced this is the right approach, so feedback is very welcome.

Compatibility Notes
--
This is in no state to be merged yet. If we decide this is the approach we want to go to, here are some notes on how to make script transition smoother.
- We could rename `get_status` to `get_stream_status`, deprecating the first, but maintaining its behaviour of polling before returning the status (in TCP).
- We should provide logically compatible enums for the old statuses in TCP and SSL
  - SSL_HOSTNAME_MISMATCH error will probably have to be equated to generic ERROR, so we should provide release notes accordingly.

### CODE REVIEW

File: core/io/stream_peer.h [hunk 1]
```diff
@@ -50,13 +50,23 @@ class StreamPeer : public Reference {
...
+
 	virtual Error put_data(const uint8_t *p_data, int p_bytes) = 0; ///< put a whole chunk of data, blocking until it sent
 	virtual Error put_partial_data(const uint8_t *p_data, int p_bytes, int &r_sent) = 0; ///< put as much data as possible, without blocking.
 
 	virtual Error get_data(uint8_t *p_buffer, int p_bytes) = 0; ///< read p_bytes of data, if p_bytes > available, it will block
 	virtual Error get_partial_data(uint8_t *p_buffer, int p_bytes, int &r_received) = 0; ///< read as much data as p_bytes into buffer, if less was read, return in r_received
 
+	virtual Status get_status() = 0; // Return the Status of this stream.
 	virtual int get_available_bytes() const = 0;
+	virtual Error poll() { return OK; } // Poll the stream looking for change in status and available bytes.
```
Comment (LikeLakers2) [inline core/io/stream_peer.h]: I know this is meant to be overridden in the specific implementations, but it seems odd to me that we're defaulting to returning OK on a method that doesn't do anything.

Comment (Faless) [reply core/io/stream_peer.h]: Well, if the Stream does not need polling and you try to poll it, there should be no error I think.
But we could also use: `ERR_UNAVAILABLE`, or better `ERR_SKIP`

Comment (Faless) [reply core/io/stream_peer.h]: To notify the user that the operation was skipped.
Yes, I guess that might work

**Comment by Faless** [comment]
Bumping to `3.1-next`, breaking changes will be implemented there,

**Comment by akien-mga** [comment]
Moving to the 4.0 milestone since that's the version where we can break compatibility. @Faless We should probably discuss this at the upcoming Godot Sprint if the changes proposed here are still relevant (note that the referenced issue is fixed already now).

**Comment by aaronfranke** [comment]
@Faless Is this still desired? If so, it needs to be rebased on the latest master branch.

If not, abandoned pull requests will be closed in the future as announced [here](https://godotengine.org/article/headsup-vulkan-merged-master-branch-unstable).

**Comment by Faless** [comment]
I didn't have time to resurrect this, but I think this is still worth working on, and 4.0 seems the right release as we are allowed to break some compatibility with older versions

**Comment by aaronfranke** [comment]
This PR has not received any new commits for almost 2 years and has not been rebased in the month after the above comments. Also, the issue this PR was supposed to fix, #4157, was fixed by #22316, so the OP also needs to be updated. Therefore, I'm closing this, it can possibly be re-opened later.

**Pull request closed without merging**